# Course 2: Python Project

# 1. Import required libraries and read the dataset.

In [7]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib.style
plt.style.use("classic")
import seaborn as sns
from sklearn.linear_model import LinearRegression 
%matplotlib inline

In [8]:
df=pd.read_csv("C:\\Users\\sahil\\Downloads\\Apps_data.csv")
dataframe = pd.DataFrame(df)
dataframe

App             Category  \
0         Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                    Coloring book moana       ART_AND_DESIGN   
2      U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                  Sketch - Draw & Paint       ART_AND_DESIGN   
4                  Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                  ...                  ...   
10836                                   Sya9a Maroc - FR               FAMILY   
10837                   Fr. Mike Schmitz Audio Teachings               FAMILY   
10838                             Parkinson Exercices FR              MEDICAL   
10839                      The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       Rating Reviews                Size     Installs  Type Price  \
0         4.1     159                 19M      10,000+  Free     0   
1         3.9     967                 14M     500,000+  Free     0   
2         4.7   87510                8.7M   5,000,000+  Free     0   
3         4.5  215644                 25M  50,000,000+  Free     0   
4         4.3     967                2.8M     100,000+  Free     0   
...       ...     ...                 ...          ...   ...   ...   
10836     4.5      38                 53M       5,000+  Free     0   
10837     5.0       4                3.6M         100+  Free     0   
10838     NaN       3                9.5M       1,000+  Free     0   
10839     4.5     114  Varies with device       1,000+  Free     0   
10840     4.5  398307                 19M  10,000,000+  Free     0   

      Content Rating                     Genres      Last Updated  \
0           Everyone               Art & Design   January 7, 2018   
1           Everyone  Art & Design;Pretend Play  January 15, 2018   
2           Everyone               Art & Design    August 1, 2018   
3               Teen               Art & Design      June 8, 2018   
4           Everyone    Art & Design;Creativity     June 20, 2018   
...              ...                        ...               ...   
10836       Everyone                  Education     July 25, 2017   
10837       Everyone                  Education      July 6, 2018   
10838       Everyone                    Medical  January 20, 2017   
10839     Mature 17+          Books & Reference  January 19, 2015   
10840       Everyone                  Lifestyle     July 25, 2018   

              Current Ver         Android Ver  
0                   1.0.0        4.0.3 and up  
1                   2.0.0        4.0.3 and up  
2                   1.2.4        4.0.3 and up  
3      Varies with device          4.2 and up  
4                     1.1          4.4 and up  
...                   ...                 ...  
10836                1.48          4.1 and up  
10837                   1          4.1 and up  
10838                   1          2.2 and up  
10839  Varies with device  Varies with device  
10840  Varies with device  Varies with device  

[10841 rows x 13 columns]

# 2. Check the first few samples, shape, info of the data and try to familiarize yourself with different features.

In [9]:
print(dataframe.shape)       # Code used to check the shape of the data set

(10841, 13)


In [10]:
dataframe.head(10)          #Code used to check the samples and top 10 rows of the data set

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   
5                         Paper flowers instructions  ART_AND_DESIGN     4.4   
6            Smoke Effect Photo Maker - Smoke Editor  ART_AND_DESIGN     3.8   
7                                   Infinite Painter  ART_AND_DESIGN     4.1   
8                               Garden Coloring Book  ART_AND_DESIGN     4.4   
9                      Kids Paint Free - Drawing Fun  ART_AND_DESIGN     4.7   

  Reviews  Size     Installs  Type Price Content Rating  \
0     159   19M      10,000+  Free     0       Everyone   
1     967   14M     500,000+  Free     0       Everyone   
2   87510  8.7M   5,000,000+  Free     0       Everyone   
3  215644   25M  50,000,000+  Free     0           Teen   
4     967  2.8M     100,000+  Free     0       Everyone   
5     167  5.6M      50,000+  Free     0       Everyone   
6     178   19M      50,000+  Free     0       Everyone   
7   36815   29M   1,000,000+  Free     0       Everyone   
8   13791   33M   1,000,000+  Free     0       Everyone   
9     121  3.1M      10,000+  Free     0       Everyone   

                      Genres        Last Updated         Current Ver  \
0               Art & Design     January 7, 2018               1.0.0   
1  Art & Design;Pretend Play    January 15, 2018               2.0.0   
2               Art & Design      August 1, 2018               1.2.4   
3               Art & Design        June 8, 2018  Varies with device   
4    Art & Design;Creativity       June 20, 2018                 1.1   
5               Art & Design      March 26, 2017                   1   
6               Art & Design      April 26, 2018                 1.1   
7               Art & Design       June 14, 2018            6.1.61.1   
8               Art & Design  September 20, 2017               2.9.2   
9    Art & Design;Creativity        July 3, 2018                 2.8   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up  
5    2.3 and up  
6  4.0.3 and up  
7    4.2 and up  
8    3.0 and up  
9  4.0.3 and up

In [11]:
dataframe.info()   # Code used to check the general info about the data set

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10841 non-null  object 
 2   Rating          9367 non-null   float64
 3   Reviews         10841 non-null  object 
 4   Size            10841 non-null  object 
 5   Installs        10841 non-null  object 
 6   Type            10840 non-null  object 
 7   Price           10841 non-null  object 
 8   Content Rating  10840 non-null  object 
 9   Genres          10841 non-null  object 
 10  Last Updated    10841 non-null  object 
 11  Current Ver     10833 non-null  object 
 12  Android Ver     10838 non-null  object 
dtypes: float64(1), object(12)
memory usage: 1.1+ MB


# 3. Check summary statistics of the dataset. List out the columns that need to be worked upon for model building.

In [12]:
dataframe.describe(include="all").T


count unique                 top   freq      mean       std  \
App              10841   9660              ROBLOX      9       NaN       NaN   
Category         10841     34              FAMILY   1972       NaN       NaN   
Rating          9367.0    NaN                 NaN    NaN  4.193338  0.537431   
Reviews          10841   6002                   0    596       NaN       NaN   
Size             10841    462  Varies with device   1695       NaN       NaN   
Installs         10841     22          1,000,000+   1579       NaN       NaN   
Type             10840      3                Free  10039       NaN       NaN   
Price            10841     93                   0  10040       NaN       NaN   
Content Rating   10840      6            Everyone   8714       NaN       NaN   
Genres           10841    120               Tools    842       NaN       NaN   
Last Updated     10841   1378      August 3, 2018    326       NaN       NaN   
Current Ver      10833   2784  Varies with device   1459       NaN       NaN   
Android Ver      10838     33          4.1 and up   2451       NaN       NaN   

                min  25%  50%  75%   max  
App             NaN  NaN  NaN  NaN   NaN  
Category        NaN  NaN  NaN  NaN   NaN  
Rating          1.0  4.0  4.3  4.5  19.0  
Reviews         NaN  NaN  NaN  NaN   NaN  
Size            NaN  NaN  NaN  NaN   NaN  
Installs        NaN  NaN  NaN  NaN   NaN  
Type            NaN  NaN  NaN  NaN   NaN  
Price           NaN  NaN  NaN  NaN   NaN  
Content Rating  NaN  NaN  NaN  NaN   NaN  
Genres          NaN  NaN  NaN  NaN   NaN  
Last Updated    NaN  NaN  NaN  NaN   NaN  
Current Ver     NaN  NaN  NaN  NaN   NaN  
Android Ver     NaN  NaN  NaN  NaN   NaN

Columns that need to be worked upon for model building:-
1)Rating
2)Reviews
3)Size
4)Installs
5)Price
6)Content rating
7)Type


# 4. Check if there are any duplicate records in the dataset? if any drop them.

In [13]:
dataframe.duplicated()    #This code is written to check for any duplicates records in the data set

dataframe.drop_duplicates(inplace=True)  #this code is written to drop the duplicates recored from the data set
dataframe.head() 


App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  Reviews  Size     Installs  Type Price Content Rating  \
0     159   19M      10,000+  Free     0       Everyone   
1     967   14M     500,000+  Free     0       Everyone   
2   87510  8.7M   5,000,000+  Free     0       Everyone   
3  215644   25M  50,000,000+  Free     0           Teen   
4     967  2.8M     100,000+  Free     0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

In [14]:
dataframe.shape

(10358, 13)

# 5. Check the unique categories of the column 'Category', Is there any invalid category? If yes, drop them.

In [15]:
unique_categories = dataframe['Category'].unique()

print(unique_categories)

['ART_AND_DESIGN' 'AUTO_AND_VEHICLES' 'BEAUTY' 'BOOKS_AND_REFERENCE'
 'BUSINESS' 'COMICS' 'COMMUNICATION' 'DATING' 'EDUCATION' 'ENTERTAINMENT'
 'EVENTS' 'FINANCE' 'FOOD_AND_DRINK' 'HEALTH_AND_FITNESS' 'HOUSE_AND_HOME'
 'LIBRARIES_AND_DEMO' 'LIFESTYLE' 'GAME' 'FAMILY' 'MEDICAL' 'SOCIAL'
 'SHOPPING' 'PHOTOGRAPHY' 'SPORTS' 'TRAVEL_AND_LOCAL' 'TOOLS'
 'PERSONALIZATION' 'PRODUCTIVITY' 'PARENTING' 'WEATHER' 'VIDEO_PLAYERS'
 'NEWS_AND_MAGAZINES' 'MAPS_AND_NAVIGATION' '1.9']


In [16]:

category_to_locate = '1.9'
rows_with_category_1_9 = dataframe[dataframe['Category'] == category_to_locate]

# Print the rows with category '1.9'
print(rows_with_category_1_9)


                                           App Category  Rating Reviews  \
10472  Life Made WI-Fi Touchscreen Photo Frame      1.9    19.0    3.0M   

         Size Installs Type     Price Content Rating             Genres  \
10472  1,000+     Free    0  Everyone            NaN  February 11, 2018   

      Last Updated Current Ver Android Ver  
10472       1.0.19  4.0 and up         NaN  


Row 10472 contains 1.9 as category however we have total of 10358 rows that we will use after removing duplicates 
hence we do not require to drop row 10472.

# 6. Check if there are missing values present in the column Rating, If any? drop them and and create a new column as 'Rating_category' by converting ratings to high and low categories(>3.5 is high rest low)

In [17]:


# Check for missing values in a particular column
column_name = 'Rating'  # Replace 'column_name' with the actual column name you want to check

missing_values = dataframe[column_name].isnull().sum()

print(f"The column '{column_name}' has {missing_values} missing values.")


The column 'Rating' has 1465 missing values.


In [18]:


# Drop rows with missing values in a Rating column
column_name = 'Rating'  # Replace 'column_name' with the actual column name
dataframe = dataframe.dropna(subset=[column_name])

# After dropping the rows, we need to reset the index
dataframe = dataframe.reset_index(drop=True)

# Print the updated DataFrame
dataframe


App             Category  \
0        Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                   Coloring book moana       ART_AND_DESIGN   
2     U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                 Sketch - Draw & Paint       ART_AND_DESIGN   
4                 Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                 ...                  ...   
8888                                      FR Calculator               FAMILY   
8889                                   Sya9a Maroc - FR               FAMILY   
8890                   Fr. Mike Schmitz Audio Teachings               FAMILY   
8891                      The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
8892      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

      Rating Reviews                Size     Installs  Type Price  \
0        4.1     159                 19M      10,000+  Free     0   
1        3.9     967                 14M     500,000+  Free     0   
2        4.7   87510                8.7M   5,000,000+  Free     0   
3        4.5  215644                 25M  50,000,000+  Free     0   
4        4.3     967                2.8M     100,000+  Free     0   
...      ...     ...                 ...          ...   ...   ...   
8888     4.0       7                2.6M         500+  Free     0   
8889     4.5      38                 53M       5,000+  Free     0   
8890     5.0       4                3.6M         100+  Free     0   
8891     4.5     114  Varies with device       1,000+  Free     0   
8892     4.5  398307                 19M  10,000,000+  Free     0   

     Content Rating                     Genres      Last Updated  \
0          Everyone               Art & Design   January 7, 2018   
1          Everyone  Art & Design;Pretend Play  January 15, 2018   
2          Everyone               Art & Design    August 1, 2018   
3              Teen               Art & Design      June 8, 2018   
4          Everyone    Art & Design;Creativity     June 20, 2018   
...             ...                        ...               ...   
8888       Everyone                  Education     June 18, 2017   
8889       Everyone                  Education     July 25, 2017   
8890       Everyone                  Education      July 6, 2018   
8891     Mature 17+          Books & Reference  January 19, 2015   
8892       Everyone                  Lifestyle     July 25, 2018   

             Current Ver         Android Ver  
0                  1.0.0        4.0.3 and up  
1                  2.0.0        4.0.3 and up  
2                  1.2.4        4.0.3 and up  
3     Varies with device          4.2 and up  
4                    1.1          4.4 and up  
...                  ...                 ...  
8888               1.0.0          4.1 and up  
8889                1.48          4.1 and up  
8890                   1          4.1 and up  
8891  Varies with device  Varies with device  
8892  Varies with device  Varies with device  

[8893 rows x 13 columns]

In [19]:
dataframe["Rating_Category"] = dataframe["Rating"].apply(lambda x: "High" if x>3.5 else "Low")
dataframe

App             Category  \
0        Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                   Coloring book moana       ART_AND_DESIGN   
2     U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                 Sketch - Draw & Paint       ART_AND_DESIGN   
4                 Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                 ...                  ...   
8888                                      FR Calculator               FAMILY   
8889                                   Sya9a Maroc - FR               FAMILY   
8890                   Fr. Mike Schmitz Audio Teachings               FAMILY   
8891                      The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
8892      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

      Rating Reviews                Size     Installs  Type Price  \
0        4.1     159                 19M      10,000+  Free     0   
1        3.9     967                 14M     500,000+  Free     0   
2        4.7   87510                8.7M   5,000,000+  Free     0   
3        4.5  215644                 25M  50,000,000+  Free     0   
4        4.3     967                2.8M     100,000+  Free     0   
...      ...     ...                 ...          ...   ...   ...   
8888     4.0       7                2.6M         500+  Free     0   
8889     4.5      38                 53M       5,000+  Free     0   
8890     5.0       4                3.6M         100+  Free     0   
8891     4.5     114  Varies with device       1,000+  Free     0   
8892     4.5  398307                 19M  10,000,000+  Free     0   

     Content Rating                     Genres      Last Updated  \
0          Everyone               Art & Design   January 7, 2018   
1          Everyone  Art & Design;Pretend Play  January 15, 2018   
2          Everyone               Art & Design    August 1, 2018   
3              Teen               Art & Design      June 8, 2018   
4          Everyone    Art & Design;Creativity     June 20, 2018   
...             ...                        ...               ...   
8888       Everyone                  Education     June 18, 2017   
8889       Everyone                  Education     July 25, 2017   
8890       Everyone                  Education      July 6, 2018   
8891     Mature 17+          Books & Reference  January 19, 2015   
8892       Everyone                  Lifestyle     July 25, 2018   

             Current Ver         Android Ver Rating_Category  
0                  1.0.0        4.0.3 and up            High  
1                  2.0.0        4.0.3 and up            High  
2                  1.2.4        4.0.3 and up            High  
3     Varies with device          4.2 and up            High  
4                    1.1          4.4 and up            High  
...                  ...                 ...             ...  
8888               1.0.0          4.1 and up            High  
8889                1.48          4.1 and up            High  
8890                   1          4.1 and up            High  
8891  Varies with device  Varies with device            High  
8892  Varies with device  Varies with device            High  

[8893 rows x 14 columns]

# 7. Check the distribution of the newly created column 'Rating_category' and comment on the distribution.

In [20]:
distribuation = dataframe["Rating_Category"].value_counts()
distribuation
#8012 is in high category that means the rating is above 3.5*
#only 880 is in low category that means the rating is below 3.5*
#Almost 90% videos has high rating and only 10% videos has a low rating

High    8013
Low      880
Name: Rating_Category, dtype: int64

# 8. Convert the column "Reviews'' to numeric data type and check the presence of outliers in the column and handle the outliers using a transformation approach.(Hint: Use log transformation)

In [21]:
dataframe['Reviews'] = pd.to_numeric(dataframe['Reviews'], errors='coerce')
dataframe.dtypes

App                 object
Category            object
Rating             float64
Reviews            float64
Size                object
Installs            object
Type                object
Price               object
Content Rating      object
Genres              object
Last Updated        object
Current Ver         object
Android Ver         object
Rating_Category     object
dtype: object

In [22]:
q1 = dataframe['Reviews'].quantile(0.25)
q3 = dataframe['Reviews'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - 1.5 * iqr
upper_bound = q3 + 1.5 * iqr

In [23]:
dataframe['Reviews'] = np.log1p(dataframe['Reviews'])
  # To handel the outliers we used log function


# 9. The column 'Size' contains alphanumeric values, treat the non numeric data and convert the column into suitable data type. (hint: Replace M with 1 million and K with 1 thousand, and drop the entries where size='Varies with device')

In [24]:
dataframe = dataframe[dataframe["Size"] != "Varies with device"]    #This code will drop all the rows which has Varies with device in the size column

dataframe.reset_index(drop=True, inplace=True)  #This code will reset the index 

dataframe

App             Category  \
0        Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                   Coloring book moana       ART_AND_DESIGN   
2     U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                 Sketch - Draw & Paint       ART_AND_DESIGN   
4                 Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                 ...                  ...   
7420                                        Chemin (fr)  BOOKS_AND_REFERENCE   
7421                                      FR Calculator               FAMILY   
7422                                   Sya9a Maroc - FR               FAMILY   
7423                   Fr. Mike Schmitz Audio Teachings               FAMILY   
7424      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

      Rating    Reviews  Size     Installs  Type Price Content Rating  \
0        4.1   5.075174   19M      10,000+  Free     0       Everyone   
1        3.9   6.875232   14M     500,000+  Free     0       Everyone   
2        4.7  11.379520  8.7M   5,000,000+  Free     0       Everyone   
3        4.5  12.281389   25M  50,000,000+  Free     0           Teen   
4        4.3   6.875232  2.8M     100,000+  Free     0       Everyone   
...      ...        ...   ...          ...   ...   ...            ...   
7420     4.8   3.806662  619k       1,000+  Free     0       Everyone   
7421     4.0   2.079442  2.6M         500+  Free     0       Everyone   
7422     4.5   3.663562   53M       5,000+  Free     0       Everyone   
7423     5.0   1.609438  3.6M         100+  Free     0       Everyone   
7424     4.5  12.894981   19M  10,000,000+  Free     0       Everyone   

                         Genres      Last Updated         Current Ver  \
0                  Art & Design   January 7, 2018               1.0.0   
1     Art & Design;Pretend Play  January 15, 2018               2.0.0   
2                  Art & Design    August 1, 2018               1.2.4   
3                  Art & Design      June 8, 2018  Varies with device   
4       Art & Design;Creativity     June 20, 2018                 1.1   
...                         ...               ...                 ...   
7420          Books & Reference    March 23, 2014                 0.8   
7421                  Education     June 18, 2017               1.0.0   
7422                  Education     July 25, 2017                1.48   
7423                  Education      July 6, 2018                   1   
7424                  Lifestyle     July 25, 2018  Varies with device   

             Android Ver Rating_Category  
0           4.0.3 and up            High  
1           4.0.3 and up            High  
2           4.0.3 and up            High  
3             4.2 and up            High  
4             4.4 and up            High  
...                  ...             ...  
7420          2.2 and up            High  
7421          4.1 and up            High  
7422          4.1 and up            High  
7423          4.1 and up            High  
7424  Varies with device            High  

[7425 rows x 14 columns]

# 10. Check the column 'Installs', treat the unwanted characters and convert the column into a suitable data type.

In [26]:
dataframe["Installs"] = dataframe["Installs"].str.replace("+", " ")
dataframe["Installs"] = dataframe["Installs"].str.replace(",", " ")

dataframe

C:\Users\sahil\AppData\Local\Temp\ipykernel_22540\2875595305.py:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  dataframe["Installs"] = dataframe["Installs"].str.replace("+", " ")
C:\Users\sahil\AppData\Local\Temp\ipykernel_22540\2875595305.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataframe["Installs"] = dataframe["Installs"].str.replace("+", " ")
C:\Users\sahil\AppData\Local\Temp\ipykernel_22540\2875595305.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in t

App             Category  \
0        Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                   Coloring book moana       ART_AND_DESIGN   
2     U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                 Sketch - Draw & Paint       ART_AND_DESIGN   
4                 Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                 ...                  ...   
7420                                        Chemin (fr)  BOOKS_AND_REFERENCE   
7421                                      FR Calculator               FAMILY   
7422                                   Sya9a Maroc - FR               FAMILY   
7423                   Fr. Mike Schmitz Audio Teachings               FAMILY   
7424      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

      Rating    Reviews  Size     Installs  Type Price Content Rating  \
0        4.1   5.075174   19M      10 000   Free     0       Everyone   
1        3.9   6.875232   14M     500 000   Free     0       Everyone   
2        4.7  11.379520  8.7M   5 000 000   Free     0       Everyone   
3        4.5  12.281389   25M  50 000 000   Free     0           Teen   
4        4.3   6.875232  2.8M     100 000   Free     0       Everyone   
...      ...        ...   ...          ...   ...   ...            ...   
7420     4.8   3.806662  619k       1 000   Free     0       Everyone   
7421     4.0   2.079442  2.6M         500   Free     0       Everyone   
7422     4.5   3.663562   53M       5 000   Free     0       Everyone   
7423     5.0   1.609438  3.6M         100   Free     0       Everyone   
7424     4.5  12.894981   19M  10 000 000   Free     0       Everyone   

                         Genres      Last Updated         Current Ver  \
0                  Art & Design   January 7, 2018               1.0.0   
1     Art & Design;Pretend Play  January 15, 2018               2.0.0   
2                  Art & Design    August 1, 2018               1.2.4   
3                  Art & Design      June 8, 2018  Varies with device   
4       Art & Design;Creativity     June 20, 2018                 1.1   
...                         ...               ...                 ...   
7420          Books & Reference    March 23, 2014                 0.8   
7421                  Education     June 18, 2017               1.0.0   
7422                  Education     July 25, 2017                1.48   
7423                  Education      July 6, 2018                   1   
7424                  Lifestyle     July 25, 2018  Varies with device   

             Android Ver Rating_Category  
0           4.0.3 and up            High  
1           4.0.3 and up            High  
2           4.0.3 and up            High  
3             4.2 and up            High  
4             4.4 and up            High  
...                  ...             ...  
7420          2.2 and up            High  
7421          4.1 and up            High  
7422          4.1 and up            High  
7423          4.1 and up            High  
7424  Varies with device            High  

[7425 rows x 14 columns]

In [27]:
dataframe['Installs'] = dataframe['Installs'].str.replace(r'\s', '', regex=True)   # This code is used to remove space between 
dataframe                                                                          #the digits in column Installs

C:\Users\sahil\AppData\Local\Temp\ipykernel_22540\1400518421.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataframe['Installs'] = dataframe['Installs'].str.replace(r'\s', '', regex=True)   # This code is used to remove space between


App             Category  \
0        Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                   Coloring book moana       ART_AND_DESIGN   
2     U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                 Sketch - Draw & Paint       ART_AND_DESIGN   
4                 Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                 ...                  ...   
7420                                        Chemin (fr)  BOOKS_AND_REFERENCE   
7421                                      FR Calculator               FAMILY   
7422                                   Sya9a Maroc - FR               FAMILY   
7423                   Fr. Mike Schmitz Audio Teachings               FAMILY   
7424      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

      Rating    Reviews  Size  Installs  Type Price Content Rating  \
0        4.1   5.075174   19M     10000  Free     0       Everyone   
1        3.9   6.875232   14M    500000  Free     0       Everyone   
2        4.7  11.379520  8.7M   5000000  Free     0       Everyone   
3        4.5  12.281389   25M  50000000  Free     0           Teen   
4        4.3   6.875232  2.8M    100000  Free     0       Everyone   
...      ...        ...   ...       ...   ...   ...            ...   
7420     4.8   3.806662  619k      1000  Free     0       Everyone   
7421     4.0   2.079442  2.6M       500  Free     0       Everyone   
7422     4.5   3.663562   53M      5000  Free     0       Everyone   
7423     5.0   1.609438  3.6M       100  Free     0       Everyone   
7424     4.5  12.894981   19M  10000000  Free     0       Everyone   

                         Genres      Last Updated         Current Ver  \
0                  Art & Design   January 7, 2018               1.0.0   
1     Art & Design;Pretend Play  January 15, 2018               2.0.0   
2                  Art & Design    August 1, 2018               1.2.4   
3                  Art & Design      June 8, 2018  Varies with device   
4       Art & Design;Creativity     June 20, 2018                 1.1   
...                         ...               ...                 ...   
7420          Books & Reference    March 23, 2014                 0.8   
7421                  Education     June 18, 2017               1.0.0   
7422                  Education     July 25, 2017                1.48   
7423                  Education      July 6, 2018                   1   
7424                  Lifestyle     July 25, 2018  Varies with device   

             Android Ver Rating_Category  
0           4.0.3 and up            High  
1           4.0.3 and up            High  
2           4.0.3 and up            High  
3             4.2 and up            High  
4             4.4 and up            High  
...                  ...             ...  
7420          2.2 and up            High  
7421          4.1 and up            High  
7422          4.1 and up            High  
7423          4.1 and up            High  
7424  Varies with device            High  

[7425 rows x 14 columns]

In [28]:
column_name = 'Installs'   #this whole code is used to check if column install has any missing value

missing_values1 = dataframe[column_name].isnull().sum()

print(f"The column '{column_name}' has {missing_values1} missing values.")

The column 'Installs' has 0 missing values.


In [29]:

non_integer_mask1 = ~dataframe['Installs'].str.isnumeric()

# Get the index numbers of rows with non-integer values
non_integer_rows1 = dataframe[non_integer_mask1].index

# Print the index numbers
print("Index numbers of rows with non-integer values in 'Installs' column:")
print(non_integer_rows1)


Index numbers of rows with non-integer values in 'Installs' column:
Int64Index([7191], dtype='int64')


In [30]:
dataframe = dataframe.drop(7191)   #To drop the row which has non integer value in column Installs

In [31]:
dataframe['Installs'] = dataframe['Installs'].astype(int)
dataframe.dtypes       #Now we can change the data type of Install column to int

App                 object
Category            object
Rating             float64
Reviews            float64
Size                object
Installs             int32
Type                object
Price               object
Content Rating      object
Genres              object
Last Updated        object
Current Ver         object
Android Ver         object
Rating_Category     object
dtype: object

# 11. Check the column 'Price' , remove the unwanted characters and convert the column into a suitable data type.

In [32]:
dataframe['Price'] = dataframe['Price'].str.replace('$', '', regex=True)
dataframe

App             Category  \
0        Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                   Coloring book moana       ART_AND_DESIGN   
2     U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                 Sketch - Draw & Paint       ART_AND_DESIGN   
4                 Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                 ...                  ...   
7420                                        Chemin (fr)  BOOKS_AND_REFERENCE   
7421                                      FR Calculator               FAMILY   
7422                                   Sya9a Maroc - FR               FAMILY   
7423                   Fr. Mike Schmitz Audio Teachings               FAMILY   
7424      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

      Rating    Reviews  Size  Installs  Type Price Content Rating  \
0        4.1   5.075174   19M     10000  Free     0       Everyone   
1        3.9   6.875232   14M    500000  Free     0       Everyone   
2        4.7  11.379520  8.7M   5000000  Free     0       Everyone   
3        4.5  12.281389   25M  50000000  Free     0           Teen   
4        4.3   6.875232  2.8M    100000  Free     0       Everyone   
...      ...        ...   ...       ...   ...   ...            ...   
7420     4.8   3.806662  619k      1000  Free     0       Everyone   
7421     4.0   2.079442  2.6M       500  Free     0       Everyone   
7422     4.5   3.663562   53M      5000  Free     0       Everyone   
7423     5.0   1.609438  3.6M       100  Free     0       Everyone   
7424     4.5  12.894981   19M  10000000  Free     0       Everyone   

                         Genres      Last Updated         Current Ver  \
0                  Art & Design   January 7, 2018               1.0.0   
1     Art & Design;Pretend Play  January 15, 2018               2.0.0   
2                  Art & Design    August 1, 2018               1.2.4   
3                  Art & Design      June 8, 2018  Varies with device   
4       Art & Design;Creativity     June 20, 2018                 1.1   
...                         ...               ...                 ...   
7420          Books & Reference    March 23, 2014                 0.8   
7421                  Education     June 18, 2017               1.0.0   
7422                  Education     July 25, 2017                1.48   
7423                  Education      July 6, 2018                   1   
7424                  Lifestyle     July 25, 2018  Varies with device   

             Android Ver Rating_Category  
0           4.0.3 and up            High  
1           4.0.3 and up            High  
2           4.0.3 and up            High  
3             4.2 and up            High  
4             4.4 and up            High  
...                  ...             ...  
7420          2.2 and up            High  
7421          4.1 and up            High  
7422          4.1 and up            High  
7423          4.1 and up            High  
7424  Varies with device            High  

[7424 rows x 14 columns]

In [33]:
column_name1 = 'Price'   #this whole code is used to check if column price has any missing value

missing_values11 = dataframe[column_name1].isnull().sum()

print(f"The column '{column_name1}' has {missing_values11} missing values.")

The column 'Price' has 0 missing values.


In [ ]:
dataframe["Price"] = dataframe["Price"].astype(float)   #here we are changing the data type of column price to float
dataframe.dtypes

# 12. Drop the columns which you think redundant for the analysis.(suggestion: drop column 'rating', since we created a new feature from it (i.e. rating_category) and the columns 'App', 'Rating' ,'Genres','Last Updated', 'Current Ver','Android Ver' columns since which are redundant for our analysis)

In [34]:
columns_to_drop =["App","Rating","Genres","Last Updated","Current Ver","Android Ver","Size"]
dataframe= dataframe.drop(columns= columns_to_drop)
dataframe

Category    Reviews  Installs  Type Price Content Rating  \
0          ART_AND_DESIGN   5.075174     10000  Free     0       Everyone   
1          ART_AND_DESIGN   6.875232    500000  Free     0       Everyone   
2          ART_AND_DESIGN  11.379520   5000000  Free     0       Everyone   
3          ART_AND_DESIGN  12.281389  50000000  Free     0           Teen   
4          ART_AND_DESIGN   6.875232    100000  Free     0       Everyone   
...                   ...        ...       ...   ...   ...            ...   
7420  BOOKS_AND_REFERENCE   3.806662      1000  Free     0       Everyone   
7421               FAMILY   2.079442       500  Free     0       Everyone   
7422               FAMILY   3.663562      5000  Free     0       Everyone   
7423               FAMILY   1.609438       100  Free     0       Everyone   
7424            LIFESTYLE  12.894981  10000000  Free     0       Everyone   

     Rating_Category  
0               High  
1               High  
2               High  
3               High  
4               High  
...              ...  
7420            High  
7421            High  
7422            High  
7423            High  
7424            High  

[7424 rows x 7 columns]

# 13. Encode the categorical columns.

In [35]:
encoded_dataframe = pd.get_dummies(dataframe, columns=["Category","Type","Content Rating"])
encoded_dataframe

Reviews  Installs Price Rating_Category  Category_ART_AND_DESIGN  \
0      5.075174     10000     0            High                        1   
1      6.875232    500000     0            High                        1   
2     11.379520   5000000     0            High                        1   
3     12.281389  50000000     0            High                        1   
4      6.875232    100000     0            High                        1   
...         ...       ...   ...             ...                      ...   
7420   3.806662      1000     0            High                        0   
7421   2.079442       500     0            High                        0   
7422   3.663562      5000     0            High                        0   
7423   1.609438       100     0            High                        0   
7424  12.894981  10000000     0            High                        0   

      Category_AUTO_AND_VEHICLES  Category_BEAUTY  \
0                              0                0   
1                              0                0   
2                              0                0   
3                              0                0   
4                              0                0   
...                          ...              ...   
7420                           0                0   
7421                           0                0   
7422                           0                0   
7423                           0                0   
7424                           0                0   

      Category_BOOKS_AND_REFERENCE  Category_BUSINESS  Category_COMICS  ...  \
0                                0                  0                0  ...   
1                                0                  0                0  ...   
2                                0                  0                0  ...   
3                                0                  0                0  ...   
4                                0                  0                0  ...   
...                            ...                ...              ...  ...   
7420                             1                  0                0  ...   
7421                             0                  0                0  ...   
7422                             0                  0                0  ...   
7423                             0                  0                0  ...   
7424                             0                  0                0  ...   

      Category_VIDEO_PLAYERS  Category_WEATHER  Type_Free  Type_Paid  \
0                          0                 0          1          0   
1                          0                 0          1          0   
2                          0                 0          1          0   
3                          0                 0          1          0   
4                          0                 0          1          0   
...                      ...               ...        ...        ...   
7420                       0                 0          1          0   
7421                       0                 0          1          0   
7422                       0                 0          1          0   
7423                       0                 0          1          0   
7424                       0                 0          1          0   

      Content Rating_Adults only 18+  Content Rating_Everyone  \
0                                  0                        1   
1                                  0                        1   
2                                  0                        1   
3                                  0                        0   
4                                  0                        1   
...                              ...                      ...   
7420                               0                        1   
7421                               0                        1   
7422                               0                       

# 14. Segregate the target and independent features (Hint: Use Rating_category as the target)

In [36]:
x= encoded_dataframe.drop("Rating_Category",axis=1)    # Here we have rmoved the Rating_Category column from the data frame and stored it in
y= encoded_dataframe["Rating_Category"]                #a Y variable for the analysis
x

Reviews  Installs Price  Category_ART_AND_DESIGN  \
0      5.075174     10000     0                        1   
1      6.875232    500000     0                        1   
2     11.379520   5000000     0                        1   
3     12.281389  50000000     0                        1   
4      6.875232    100000     0                        1   
...         ...       ...   ...                      ...   
7420   3.806662      1000     0                        0   
7421   2.079442       500     0                        0   
7422   3.663562      5000     0                        0   
7423   1.609438       100     0                        0   
7424  12.894981  10000000     0                        0   

      Category_AUTO_AND_VEHICLES  Category_BEAUTY  \
0                              0                0   
1                              0                0   
2                              0                0   
3                              0                0   
4                              0                0   
...                          ...              ...   
7420                           0                0   
7421                           0                0   
7422                           0                0   
7423                           0                0   
7424                           0                0   

      Category_BOOKS_AND_REFERENCE  Category_BUSINESS  Category_COMICS  \
0                                0                  0                0   
1                                0                  0                0   
2                                0                  0                0   
3                                0                  0                0   
4                                0                  0                0   
...                            ...                ...              ...   
7420                             1                  0                0   
7421                             0                  0                0   
7422                             0                  0                0   
7423                             0                  0                0   
7424                             0                  0                0   

      Category_COMMUNICATION  ...  Category_VIDEO_PLAYERS  Category_WEATHER  \
0                          0  ...                       0                 0   
1                          0  ...                       0                 0   
2                          0  ...                       0                 0   
3                          0  ...                       0                 0   
4                          0  ...                       0                 0   
...                      ...  ...                     ...               ...   
7420                       0  ...                       0                 0   
7421                       0  ...                       0                 0   
7422                       0  ...                       0                 0   
7423                       0  ...                       0                 0   
7424                       0  ...                       0                 0   

      Type_Free  Type_Paid  Content Rating_Adults only 18+  \
0             1          0                               0   
1             1          0                               0   
2             1          0                               0   
3             1          0                               0   
4             1          0                               0   
...         ...        ...                             ...   
7420          1          0                               0   
7421          1          0                               0   
7422          1          0                               0   
7423          1          0                               0   
7424          1          0                               0   

      Content Rating_Everyone  Content Rating_Everyone 10+  \
0    

In [37]:
y    # Here is the Rating_category variable

0       High
1       High
2       High
3       High
4       High
        ... 
7420    High
7421    High
7422    High
7423    High
7424    High
Name: Rating_Category, Length: 7424, dtype: object

# 15. Split the dataset into train and test.

In [39]:
from sklearn.model_selection import train_test_split


# Split the dataset into train and test sets
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.3, random_state=42)

# Print the shapes of the train and test sets
print("Train set shape:", x_train.shape, y_train.shape)
print("Test set shape:", x_test.shape, y_test.shape)


Train set shape: (5196, 44) (5196,)
Test set shape: (2228, 44) (2228,)
